controller/backup_controller.go

```diff
@@ -434,99 +510,139 @@ func (bc *BackupController) backupCreation(log logrus.FieldLogger, engineClient
-			state = types.BackupStateCompleted
-			event(nil, state, backup, volume)
-
-			syncTime := time.Now().UTC()
-			backupVolume, err := bc.ds.GetBackupVolume(volumeName)
-			if err == nil {
-				// Request backup_volume_controller to reconcile BackupVolume immediately.
-				backupVolume.Spec.SyncRequestedAt = syncTime
-				if _, err = bc.ds.UpdateBackupVolume(backupVolume); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
-					log.WithError(err).Errorf("Error updating backup volume %s spec", volume.Name)
-				}
-			} else if err != nil && apierrors.IsNotFound(err) {
-				// Request backup_target_controller to reconcile BackupTarget immediately.
-				backupTarget, err := bc.ds.GetBackupTarget(types.DefaultBackupTargetName)
-				if err != nil {
-					log.WithError(err).Warn("Failed to get backup target")
-					return
-				}
-				backupTarget.Spec.SyncRequestedAt = syncTime
-				if _, err = bc.ds.UpdateBackupTarget(backupTarget); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
-					log.WithError(err).Warn("Failed to update backup target")
+	lock  sync.Mutex
+	stale bool
+
+	syncCallback func(key string)
+
+	ctx  context.Context
+	quit context.CancelFunc
+}
+
+func NewBackupMonitor(logger logrus.FieldLogger, backup *longhorn.Backup, engineClient engineapi.EngineClient, syncCallback func(key string)) *BackupMonitor {
+	ctx, quit := context.WithCancel(context.Background())
+	monitor := &BackupMonitor{
+		logger: logger,
+
+		namespace:          backup.Namespace,
+		backupName:         backup.Name,
+		snapshotName:       backup.Spec.SnapshotName,
+		engineClient:       engineClient,
+		engineBackupStatus: &types.EngineBackupStatus{},
+
+		lock:  sync.Mutex{},
+		stale: true,
+
+		syncCallback: syncCallback,
+
+		ctx:  ctx,
+		quit: quit,
+	}
+
+	go monitor.monitorBackups()
+
+	return monitor
+}
+
+func (m *BackupMonitor) Close() {
+	m.quit()
+}
+
+func (m *BackupMonitor) syncBackups() error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	engineBackupStatusList, err := m.engineClient.SnapshotBackupStatus()
+	if err != nil {
+		return err
+	}
+
+	engineBackupStatus, ok := engineBackupStatusList[m.backupName]
+	if !ok || engineBackupStatus == nil || engineBackupStatus.SnapshotName != m.snapshotName {
+		return fmt.Errorf("cannot find matched snapshot/backup %s/%s in longhorn engine", m.snapshotName, m.backupName)
+	}
+
+	// new information, request a resync for this backup
+	if !reflect.DeepEqual(m.engineBackupStatus, engineBackupStatus) || m.stale {
+		m.engineBackupStatus = engineBackupStatus
+		m.stale = false
+		key := m.namespace + "/" + m.backupName
+		m.syncCallback(key)
+	}
+	return nil
+}
+
+func (m *BackupMonitor) monitorBackups() {
+	ticker := time.NewTicker(BackupMonitorTickPeriod)
+	defer ticker.Stop()
+	timeSinceLastSync := time.Duration(0)
+	for {
+		select {
+		case <-m.ctx.Done():
+			return
+		case <-ticker.C:
+			timeSinceLastSync += BackupMonitorTickPeriod
+			if timeSinceLastSync > BackupMonitorSyncPeriod || m.stale {
```

Comment:
NIT: this should be >= syncPeriod otherwise you will only sync on the 3rd tick instead of the 2nd tick.